core-java/src/test/java/com/baeldung/deepcopy/DeepCopyTest.java

```diff
@@ -0,0 +1,131 @@
+        User pm = new User("Prime", "Minister", address);
+        User deepCopy = new User(pm);
+
+        assertNotSame(deepCopy, pm);
+    }
+
+    @Test
+    public void whenModifyingOriginalObject_ThenConstructorCopyShouldNotChange() {
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+        User deepCopy = new User(pm);
+
+        address.setCountry("Great Britain");
+
+        assertThat(deepCopy.getAddress().getCountry())
+                .isNotEqualTo(pm.getAddress().getCountry());
+    }
+
+    @Test
+    public void whenModifyingOriginalObject_ThenCloneCopyShouldNotChange() {
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+        User deepCopy = (User) pm.clone();
+
+        address.setCountry("Great Britain");
+
+        assertThat(deepCopy.getAddress().getCountry())
+                .isNotEqualTo(pm.getAddress().getCountry());
+    }
+
+    @Test
+    public void whenModifyingOriginalObject_thenCommonsCloneShouldNotChange() {
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+        User deepCopy = (User) SerializationUtils.clone(pm);
+
+        address.setCountry("Great Britain");
+
+        assertThat(deepCopy.getAddress().getCountry())
+                .isNotEqualTo(pm.getAddress().getCountry());
+    }
+
+    @Test
+    public void whenModifyingOriginalObject_thenGsonCloneShouldNotChange() {
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+        Gson gson = new Gson();
+        User deepCopy = gson.fromJson(gson.toJson(pm), User.class);
+
+        address.setCountry("Great Britain");
+
+        assertThat(deepCopy.getAddress().getCountry())
+                .isNotEqualTo(pm.getAddress().getCountry());
+    }
+
+    @Test
+    public void whenModifyingOriginalObject_ThenJacksonCopyShouldNotChange() throws IOException {
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+        ObjectMapper objectMapper = new ObjectMapper();
+        User deepCopy = objectMapper.readValue(objectMapper.writeValueAsString(pm), User.class);
+
+        address.setCountry("Great Britain");
+
+        assertThat(deepCopy.getAddress().getCountry())
+                .isNotEqualTo(pm.getAddress().getCountry());
+    }
+
+
+    @Test
+    public void shouldDisplayHowLongEachMethodTake() throws CloneNotSupportedException, IOException {
+        Address address = new Address("Downing St 10", "London", "England");
+        User pm = new User("Prime", "Minister", address);
+
+        long start = System.currentTimeMillis();
+        for (int i = 0; i < 1000000; i++) {
```

Comment:
Two notes: First, wouldn't it be good to just use a constant here? And second, one million is kind of a lot for it to be in a unit test that gets run automatically during the build, so let's add @Ignore to this test. If someone wants to run it manually, they can just remove the annotation.

Reply:
Applied